include/core/Layer.hpp

```diff
@@ -0,0 +1,141 @@
+#endif
+
+namespace ngen
+{
+
+    class Layer
+    {    
+        public:
+
+        #ifdef NGEN_MPI_ACTIVE
+            typedef hy_features::HY_Features_MPI feature_type;
+        #else
+            typedef hy_features::HY_Features feature_type;
+        #endif
+
+        Layer(
+                const LayerDescription& desc, 
+                const std::vector<std::string>& p_u, 
+                const Simulation_Time& s_t, 
+                feature_type& f, 
+                geojson::GeoJSON cd, 
+                long idx) :
+            description(desc),
+            processing_units(p_u),
+            simulation_time(s_t),
+            features(f),
+            catchment_data(cd),
+            output_time_index(idx)
+        {
+
+        }
+
+        virtual ~Layer() {}
+
+        /***
+         * @brief Return the next timestep that will be processed by this layer in epoc time units
+        */
+        time_t next_timestep_epoch_time() { return simulation_time.next_timestep_epoch_time(); }
+
+
+        /***
+         * @brief Return the last timesteo that was processed by this layer in epoc time units
+        */
+        time_t current_timestep_epoch_time() { return simulation_time.get_current_epoc_time(); }
+
+
+        /***
+         * @brief Return the numeric id of this layer
+        */
+        int get_id() const { return this->description.id; }
+
+        /***
+         * @brief Return the name of this layer
+        */
+        const std::string& get_name() const { return this->description.name; }
+
+        /***
+         * @brief Return this time_step interval used for this layer
+        */
+        const double get_time_step() const { return this->description.time_step; }
+
+        /***
+         * @brief Return the units for the time_step value of this layer
+        */
+        const std::string& get_time_step_units() const { return this->description.time_step_units; }
+
+        /***
+         * @breif Run one simulation timestep for each model in this layer
+        */
+        virtual void update_models()
+        {
+            auto idx = simulation_time.next_timestep_index();
+            auto step = simulation_time.get_output_interval_seconds();
+            
+            //std::cout<<"Output Time Index: "<<output_time_index<<std::endl;
+            if(output_time_index%100 == 0) std::cout<<"Running timestep " << output_time_index <<std::endl;
+            std::string current_timestamp = simulation_time.get_timestamp(output_time_index);
+            for(const auto& id : processing_units) 
+            {
+                int sub_time = output_time_index;
+                //std::cout<<"Running cat "<<id<<std::endl;
+                auto r = features.catchment_at(id);
+                //TODO redesign to avoid this cast
+                auto r_c = std::dynamic_pointer_cast<realization::Catchment_Formulation>(r);
+                double response = r_c->get_response(output_time_index, description.time_step);
```

Comment:
will `description.time_step` and `simulation_time.get_output_interval_seconds()` ever differ? Are they redundant or is there a current/potential future semantic difference?

Reply:
Actually since it looks like `description` has a `time_step_units`, it's likely that `description.time_step` should not be used here and `simulation_time.get_output_interval_seconds()` should be used instead, or that `description.time_step` should be put through a unit conversion first (probably the former).